### ISSUE TITLE
Does geopy's geocode feature support caching?

### ISSUE THREAD

**Issue by kenseii** [geopy/geopy#375]
Hello, i tried to look up in the docs and saw that cache isnt mentioned anywhere, i was wondering if geopy cant use a caching feature to hold geocode responses and in case there is a repetitive/recurring text to be geocoded, it can fetch it from the existing cache; saving both time and api requests.
It can be an option maybe `geo.geocode("something", use_cache=True)`

Let me know if it is fine to contribute such a function/or if it exist already.

Thanks

**Comment by KostyaEsmukov** [comment]
No, there's no caching built in.

This is mentioned in the project's roadmap: #267 , and I believe this would make a great complementary module for geopy (in the same fashion as [RateLimiter](https://geopy.readthedocs.io/en/1.20.0/#geopy.extra.rate_limiter.RateLimiter)).

The implementation might be tricky though.

For reverse geocoding `geohash` algorithm #324 could be used to reduce the caching space.

For forward geocoding this is much trickier. A simple LRU cache is too easy to implement for it to have a reason to be included to geopy.

That said, if you need to solve that problem fast, roll up your own caching. If you're willing to reason about the proper implementation inside geopy, I'm open to discussion.

**Comment by KostyaEsmukov** [comment]
Just for the sake of completeness of my reply: by simple LRU cache I mean the following:

```pycon
In [1]: import functools
   ...: from geopy.geocoders import Nominatim
   ...:
   ...: geolocator = Nominatim(user_agent="specify_your_app_name_here")
   ...: geocode = functools.lru_cache(maxsize=128)(functools.partial(geolocator.geocode, timeout=5))

In [2]: %%time
   ...: geocode("London")
   ...:
   ...:
CPU times: user 17.2 ms, sys: 3.06 ms, total: 20.2 ms
Wall time: 232 ms
Out[2]: Location(London, Greater London, England, SW1A 2DX, United Kingdom, (51.5073219, -0.1276474, 0.0))

In [3]: %%time
   ...: geocode("London")
   ...:
   ...:
CPU times: user 3 µs, sys: 1e+03 ns, total: 4 µs
Wall time: 5.96 µs
Out[3]: Location(London, Greater London, England, SW1A 2DX, United Kingdom, (51.5073219, -0.1276474, 0.0))
```

**Comment by kenseii** [comment]
what about a caching method like this [one](https://github.com/gboeing/osmnx/blob/bdfb49c3d7f43613df7542b7a1e456776856821c/osmnx/core.py#L67)

**Comment by KostyaEsmukov** [comment]
Caching to file should not be the only storage backend. Other typical caching backends also include local memory (like that LRU cache sample) and redis/memcached.

**Comment by alessandro-gentilini** [comment]
Since I am using Python 2.7 I miss the `@cache` and so here a quick and dirty implementation of cache persisting on file.

```
from geopy.geocoders import Nominatim
import time
import pickle
from pathlib import Path

def query_geocoder_server(query):
    time.sleep(1) # https://operations.osmfoundation.org/policies/nominatim/
    return geolocator.reverse(query).address

def get_address(query):
    if query not in cache:
        cache[query] = query_geocoder_server(query)
    return cache[query]

cache_path = 'cache.p'

cache = dict()
if Path(cache_path).exists():
    cache = pickle.load(open(cache_path, 'rb'))

geolocator = Nominatim(user_agent="Alessandro")  

q1='44.313611 11.583889'
q2='44.329444 11.587778'

print(get_address(q1))
print(get_address(q2))

print(get_address(q1))
print(get_address(q2))

pickle.dump(cache, open(cache_path, 'wb'))
```

**Comment by CharString** [comment]
I bumped into this whilst answering [a question on StackOverflow](https://stackoverflow.com/a/68727013/383793)

A shame you can't inject your own requests Session here:
https://github.com/geopy/geopy/blob/ade9c1b68c83a1fc76d90bc0ff603f6b34dfbbbf/geopy/adapters.py#L396

But this, before creating the `Nominatim` seems to work:
```
import requests_cache
requests_cache.install_cache('geopy_cache')
```

**Comment by JWCook** [comment]
I agree with the idea of using requests-cache here, although I'm a bit biased. :grin: 

It does support Redis, among several other backends. And I'd also agree that the best way to use it with geopy would be to add support for passing your own session object, so users could use `requests_cache.CachedSession` instead of `requests_cache.install_cache`. [The difference is explained here](https://requests-cache.readthedocs.io/en/stable/user_guide.html#general-usage).

**Comment by KostyaEsmukov** [comment]
> A shame you can't inject your own requests Session here:

I don't think this is a good idea to include API tokens in caches, especially when using an external storage such as Redis for that.

Caching should be done on a higher level than Adapters.

**Comment by JWCook** [comment]
requests-cache works at the session level (`Session.send()`), not the adapter level, and also has an option for [excluding API tokens and other params](https://requests-cache.readthedocs.io/en/v0.7.3/user_guide.html#request-parameters).

**Comment by KostyaEsmukov** [comment]
I didn't mean requests' Adapters. I meant [geopy Adapters](https://geopy.readthedocs.io/en/stable/#module-geopy.adapters) (which, BTW, have implementations for urllib and aiohttp rather than just requests).

Also that option for excluding API tokens does require to look inside a specific geocoder implementation to find out how it passes the tokens. Needless to say this wouldn't work with geocoders other than the one you've looked at. If you choose that route (a single geocoder and a single Adapter) I guess it would be much easier to not use geopy completely and work with the API of the target service directly.

**Comment by KostyaEsmukov** [comment]
Good and actually working workarounds have already been given in this issue twice:
- https://github.com/geopy/geopy/issues/375#issuecomment-542084796
- https://github.com/geopy/geopy/issues/375#issuecomment-752019669

**Comment by patvdleer** [comment]
> I bumped into this whilst answering [a question on StackOverflow](https://stackoverflow.com/a/68727013/383793)
> 
> A shame you can't inject your own requests Session here:
> 
> [geopy/geopy/adapters.py](https://github.com/geopy/geopy/blob/ade9c1b68c83a1fc76d90bc0ff603f6b34dfbbbf/geopy/adapters.py#L396)
> 
> Line 396 in [ade9c1b](/geopy/geopy/commit/ade9c1b68c83a1fc76d90bc0ff603f6b34dfbbbf)
>  self.session = requests.Session() 
> 
> But this, before creating the `Nominatim` seems to work:
> 
> ```
> import requests_cache
> requests_cache.install_cache('geopy_cache')
> ```

I tried the but with [Betamax](https://github.com/betamaxpy/betamax), so far no luck.

```python
geolocator = Nominatim(user_agent="MyApp")
with Betamax(geolocator.adapter.session) as vcr:
    # Do request here
    pass
```

Sadly this raises a `BetamaxError` error with `"A request was made that could not be handled."`
